{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9fbbad6-5f66-4abf-8cf9-18831bd9cb16",
   "metadata": {},
   "outputs": [],
   "source": [
    "# scRNA analysis trial of the metabolism focused  scCellfie package\n",
    "#Description: a metabolic potential analysis of the intestine - focusing no enterocytes\n",
    "#Figures generated Figure 5A, Figure 5B, Figure 5C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f01c253c-2dc9-4b18-889d-c3f094a1134d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## To avoid warnings\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "#Load the packages\n",
    "import sccellfie\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import glasbey\n",
    "\n",
    "import textwrap\n",
    "\n",
    "print(sc.__version__)\n",
    "print(sccellfie.__version__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3cca8b9d-6ae9-4409-92e4-bfc239949c41",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load the data \n",
    "# An asterisk beside the chunk means it is still running \n",
    "adata = sc.read(filename='/path/to/analysis/directory/Seurat_Files/Analysis2.7_Ileum_counts_scCellfie.h5ad')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed3baf8f-b788-43f2-b14c-3b0e45b01b96",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(adata.obs.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4583136-ee20-4d32-a1d7-11e6d4233bea",
   "metadata": {},
   "outputs": [],
   "source": [
    "#once loaded check the contents of the file\n",
    "adata \n",
    "print(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ae34551-e718-4a98-b8c1-11ae77954978",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Specify the batch \n",
    "batch_key = 'Mouse' "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e81d7d78-af5e-4a29-9d13-7521a0418d20",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Basic scCllfie Pipeline\n",
    "results = sccellfie.run_sccellfie_pipeline(adata,\n",
    "                                           organism='mouse',\n",
    "                                           sccellfie_data_folder=None,\n",
    "                                           n_counts_col=None, # Total counts per cell will be computed if left as None,\n",
    "                                           process_by_group=False, # Whether to do the processing by cell groups\n",
    "                                           groupby=None, # Column indicating cell groups if `process_by_group=True`\n",
    "                                           neighbors_key='neighbors', # Neighbors information if precomputed. Otherwise, it will be computed here\n",
    "                                           n_neighbors=10, # Number of neighbors to use\n",
    "                                           batch_key=batch_key, # The appropriate batch to correct for\n",
    "                                           threshold_key='sccellfie_threshold',  # This is for using the default database. If personalized thresholds are used, specificy column name\n",
    "                                           smooth_cells=True, # Whether to perform gene expression smoothing before running the tool\n",
    "                                           alpha=0.33, # Importance of neighbors' expression for the smoothing (0 to 1)\n",
    "                                           chunk_size=5000, # Number of chunks to run the processing steps (helps with the memory)\n",
    "                                           disable_pbar=False,\n",
    "                                           save_folder=\"/path/to/analysis/directory/scCellfie/results/\", # In case results will be saved. If so, results will not be returned and should be loaded from the folder (see sccellfie.io.load_data function\n",
    "                                           save_filename=\"scCell_fie_analyzed\" # Name for saving the files, otherwise a default name will be used\n",
    "                                          )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c0110db-7d45-4817-b6e5-e58ab4b08f69",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Can I view the results \n",
    "results.keys()\n",
    "print(results['task_info'].head())\n",
    "print(results['task_info'].columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57c26712-1e96-4be3-b03b-dc8ab92a4fdf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Save a different way \n",
    "sccellfie.io.save_adata(adata=results['adata'], output_directory='/path/to/analysis/directory/scCellfie/results/', filename='Ileum_scCellfie')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "213028ca-811a-44dd-a06e-8319d31237db",
   "metadata": {},
   "outputs": [],
   "source": [
    "sccellfie.io.save_adata?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80919b0b-5c9f-42a1-a0f9-ce2a9adbe11d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Save the Task_info which is not usually stored correctly\n",
    "import pandas as pd\n",
    "\n",
    "#save it explicitly:\n",
    "if 'task_info' in results:\n",
    "    task_info_df = results['task_info']\n",
    "    task_info_df.to_csv(\"/path/to/analysis/directory/scCellfie/results/scCell_fie_task_info.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e019a831-cdf2-4d25-bb7b-63f5b059188d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load results from saved \n",
    "import sccellfie.io\n",
    "\n",
    "print(hasattr(sccellfie.io, \"load_data\")) # Should return True if the function exists\n",
    "help(sccellfie.io.load_data)\n",
    "\n",
    "adata = sccellfie.io.load_adata(folder = '/path/to/analysis/directory/scCellfie/results/', filename = 'Ileum_scCellfie', \n",
    "                        reactions_filename='Ileum_scCellfie_reactions', metabolic_tasks_filename='Ileum_scCellfie_metabolic_tasks',  verbose=True)\n",
    "\n",
    "# Load the Results task_info\n",
    "task_info_df = pd.read_csv(\"/path/to/analysis/directory/scCellfie/results/scCell_fie_task_info.csv\", index_col=0)\n",
    "\n",
    "#Assign Task_info\n",
    "adata.metabolic_tasks.uns['task_info'] = task_info_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55aaee84-81d4-4a08-8817-72edb9cebff5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Visualize results part 1 \n",
    "cell_group = 'secondlevel'\n",
    "infections = \"InfectionStatus\"\n",
    "palette = glasbey.extend_palette('Set2',\n",
    "    palette_size=max([10, adata.metabolic_tasks.obs[cell_group].nunique()])\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ae5d350-d506-4635-97e0-f0d30d054cd4",
   "metadata": {},
   "outputs": [],
   "source": [
    "for task in adata.metabolic_tasks.var_names:\n",
    "    print(task)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2d5c277-47c1-4db8-a561-450fb8f90b28",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Exploration \n",
    "metabolic_tasks = ['Glycogen biosynthesis',\n",
    "                   'ATP generation from glucose (hypoxic conditions) - glycolysis',\n",
    "                   'Synthesis of taurine from cysteine',\n",
    "                   'Synthesis of serotonin from tryptophan',\n",
    "                   \"Synthesis of arginine from glutamine\",\n",
    "                   \"Reactive oxygen species detoxification (H2O2 to H2O)\",\n",
    "                   \"Phosphatidyl-choline synthesis\",\n",
    "                   \"Oxidative phosphorylation via succinate-coenzyme Q oxidoreductase (COMPLEX II)\",\n",
    "                   \"O-linked glycosylation\",\n",
    "                   \"N-linked glycosylation\",\n",
    "                   \"Krebs cycle - NADH generation\",\n",
    "                   \"Glycogen degradation\",\n",
    "                   \"Mannose degradation (to fructose-6-phosphate)\",\n",
    "                   \"Cholesterol synthesis\",\n",
    "                   \"Calnexin/calreticulin cycle\"\n",
    "                  ]\n",
    "\n",
    "plt.rcParams['figure.figsize'] = (3,3)\n",
    "plt.rcParams['font.size'] = 10\n",
    "\n",
    "sc.pl.embedding(adata.metabolic_tasks,\n",
    "                color=[cell_group] + metabolic_tasks,\n",
    "                ncols=1,\n",
    "                palette=palette,\n",
    "                frameon=False,\n",
    "                basis='X_wnn.umap_cc',\n",
    "                wspace=0.7,\n",
    "                title=[\"\\n\".join(textwrap.wrap(t, width=60)) for t in [cell_group] + metabolic_tasks],\n",
    "                cmap='OrRd'\n",
    "               )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eeb5e69e-b870-4aaa-958f-5e13a92cb245",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = sccellfie.plotting.create_multi_violin_plots(adata.metabolic_tasks,\n",
    "                                                         features=metabolic_tasks,\n",
    "                                                         groupby=cell_group, \n",
    "                                                         stripplot=False,\n",
    "                                                         n_cols=2,\n",
    "                                                         ylabel='Metabolic Score'\n",
    "                                                        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78206acf-0ff5-43b7-a558-a51fd40ac157",
   "metadata": {},
   "outputs": [],
   "source": [
    "#stacked violins\n",
    "ax = sc.pl.stacked_violin(adata.metabolic_tasks, metabolic_tasks, groupby=cell_group, swap_axes=True, dendrogram=False, standard_scale='var')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "559586bc-173c-49c5-a85d-2bfe11eeced5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Heatmap \n",
    "ax = sc.pl.heatmap(adata.metabolic_tasks, var_names=metabolic_tasks, groupby=cell_group, cmap=\"YlGnBu\", swap_axes=True, dendrogram=True,\n",
    "                   figsize=(16, 4)\n",
    "                  )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c308a9b-718f-485c-a20c-5e69dec80240",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Aggregate based on cell type \n",
    "agg = sccellfie.expression.aggregation.agg_expression_cells(adata.metabolic_tasks, groupby=cell_group, agg_func='trimean')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d45906d-6c01-49c2-a82b-238a5b0df81d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#min max normalization \n",
    "input_df = sccellfie.preprocessing.matrix_utils.min_max_normalization(agg.T, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f86883df-6da6-4465-95ce-df5ceb69326b",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(16, 4))\n",
    "g = sns.heatmap(input_df.loc[metabolic_tasks,:], cmap='YlGnBu', linewidths=0.5, xticklabels=1, yticklabels=1)\n",
    "\n",
    "cbar = g.collections[0].colorbar\n",
    "cbar.set_label('Scaled metabolic activity', size=14, rotation=270, labelpad=25)  # Change colorbar label size and rotation\n",
    "\n",
    "# Uncomment code below to save figure\n",
    "# plt.savefig('./figures/Heatmap-Seaborn.pdf', dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d4457d3-7d9b-430f-b755-e7879e186083",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Radial Plots \n",
    "df_melted = pd.melt(input_df.reset_index(), id_vars='Task', var_name='cell_type', value_name='scaled_trimean')\n",
    "df_melted = df_melted.rename(columns={'Task': 'metabolic_task'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27056fdf-c35d-4021-98d7-6657f1213362",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_types = ['Enterocytes', 'Stem Cells', 'Transit Amplifying Cells', 'Goblet Cells']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3436c134-95f6-4278-bc59-2a3886a5e9bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create figure with subplots\n",
    "\n",
    "fig = plt.figure(figsize=(16, 16))\n",
    "ax1 = fig.add_subplot(221, projection='polar')\n",
    "ax2 = fig.add_subplot(222, projection='polar')\n",
    "ax3 = fig.add_subplot(223, projection='polar')\n",
    "ax4 = fig.add_subplot(224, projection='polar')\n",
    "\n",
    "for i, (cell, ax) in enumerate(zip(cell_types, [ax1, ax2, ax3, ax4])):\n",
    "    sccellfie.plotting.create_radial_plot(df_melted,\n",
    "                                          adata.metabolic_tasks.uns['task_info'],\n",
    "                                          cell_type=cell,\n",
    "                                          ax=ax,\n",
    "                                          show_legend=i == 1,\n",
    "                                          ylim=1.0,\n",
    "                                          sort_by_value=True\n",
    "                                          )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dee41409-75da-4696-ae27-3d69a1aa59b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "\n",
    "# Create group label for disease x celltype\n",
    "adata.metabolic_tasks.obs['group_label'] = (\n",
    "    adata.metabolic_tasks.obs['secondlevel'].astype(str) + \"_\" +\n",
    "    adata.metabolic_tasks.obs['InfectionStatus'].astype(str)\n",
    ")\n",
    "\n",
    "# Aggregate using group_label\n",
    "agg = sccellfie.expression.aggregation.agg_expression_cells(\n",
    "    adata.metabolic_tasks,\n",
    "    groupby='group_label',\n",
    "    agg_func='trimean'\n",
    ")\n",
    "\n",
    "# Normalize and melt for plotting\n",
    "input_df = sccellfie.preprocessing.matrix_utils.min_max_normalization(agg.T, axis=1)\n",
    "df_melted = pd.melt(input_df.reset_index(), id_vars='Task', var_name='group_label', value_name='scaled_trimean')\n",
    "df_melted = df_melted.rename(columns={'Task': 'metabolic_task'})\n",
    "\n",
    "\n",
    "df_melted['metabolic_task'] = df_melted['metabolic_task'].astype(str).str.strip()\n",
    "adata.metabolic_tasks.uns['task_info']['Task'] = adata.metabolic_tasks.uns['task_info']['Task'].astype(str).str.strip()\n",
    "\n",
    "# Filter valid tasks\n",
    "valid_tasks = set(adata.metabolic_tasks.uns['task_info']['Task'])\n",
    "df_melted = df_melted[df_melted['metabolic_task'].isin(valid_tasks)].copy()\n",
    "\n",
    "# Filter to enterocytes\n",
    "enterocyte_df = df_melted[df_melted['group_label'].str.startswith(\"Enterocyte\")].copy()\n",
    "\n",
    "# Extract condition and cell type\n",
    "enterocyte_df['condition'] = enterocyte_df['group_label'].str.split('_').str[1]\n",
    "enterocyte_df['cell_type'] = enterocyte_df['group_label'].str.split('_').str[0]\n",
    "\n",
    "\n",
    "conditions = enterocyte_df['condition'].unique()\n",
    "\n",
    "n = len(conditions)\n",
    "ncols = 3\n",
    "nrows = (n // ncols) + int(n % ncols != 0) + 1  # +1 for legend\n",
    "fig, axes = plt.subplots(nrows=nrows, ncols=ncols, subplot_kw={'projection': 'polar'}, figsize=(5 * ncols, 5 * nrows))\n",
    "axes = axes.flatten()\n",
    "\n",
    "# Plot each condition\n",
    "for i, (condition, ax) in enumerate(zip(conditions, axes)):\n",
    "    condition_df = enterocyte_df[enterocyte_df['condition'] == condition]\n",
    "    sccellfie.plotting.create_radial_plot(\n",
    "        condition_df,\n",
    "        adata.metabolic_tasks.uns['task_info'],\n",
    "        ax=ax,\n",
    "        show_legend=False,\n",
    "        ylim=1.0,\n",
    "        sort_by_value=True\n",
    "    )\n",
    "    ax.set_title(condition)\n",
    "\n",
    "# Legend subplot\n",
    "legend_index = len(conditions)\n",
    "if legend_index < len(axes):\n",
    "    legend_ax = axes[legend_index]\n",
    "    legend_ax.set_axis_off()\n",
    "\n",
    "    # Dummy plot for legend handles\n",
    "    dummy_ax = plt.subplot(111, polar=True)\n",
    "    condition_df = enterocyte_df[enterocyte_df['condition'] == conditions[0]]\n",
    "    tmp = sccellfie.plotting.create_radial_plot(\n",
    "        condition_df,\n",
    "        adata.metabolic_tasks.uns['task_info'],\n",
    "        ax=dummy_ax,\n",
    "        show_legend=True,\n",
    "        ylim=1.0,\n",
    "        sort_by_value=True\n",
    "    )\n",
    "    handles, labels = dummy_ax.get_legend_handles_labels()\n",
    "    plt.close(dummy_ax.figure)\n",
    "\n",
    "    # Add legend\n",
    "    legend_ax.legend(handles, labels, loc='center', frameon=False)\n",
    "    legend_ax.set_title(\"Metabolic Systems\")\n",
    "\n",
    "\n",
    "for j in range(legend_index + 1, len(axes)):\n",
    "    fig.delaxes(axes[j])\n",
    "\n",
    "plt.tight_layout()\n",
    "# Save figure\n",
    "# Figure 5A\n",
    "output_path = \"/path/to/analysis/directory/scCellfie/Images/scCellfie_radial_plots_enterocytes.svg\"\n",
    "fig.savefig(output_path, format=\"svg\", dpi=300)\n",
    "plt.show()\n",
    "print(f\"Figure saved to: {output_path}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d988f7d-6047-4793-9792-92016f2adacf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check the main structure\n",
    "print(adata.obs.head())\n",
    "\n",
    "task_matrix = adata.metabolic_tasks.X  # this is usually a numpy array\n",
    "\n",
    "# Get cell names\n",
    "cells = adata.metabolic_tasks.obs_names\n",
    "\n",
    "# Get task names\n",
    "tasks = adata.metabolic_tasks.var_names\n",
    "\n",
    "# Convert to DataFrame\n",
    "tasks_df = pd.DataFrame(task_matrix, index=cells, columns=tasks)\n",
    "\n",
    "# Preview\n",
    "tasks_df.head()\n",
    "tasks_df.to_csv(\"/path/to/analysis/directory/scCellfie/cell_task_scores.csv\", index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ef3ca39-396a-405e-802a-dcb761e3eb20",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For the Nippo MEtabolism Lipids - Which of the specific pathways are the two higest \n",
    "# Filter and sort again (if not done yet)\n",
    "lipid_tasks = adata.metabolic_tasks.uns['task_info']\n",
    "lipid_tasks = lipid_tasks[lipid_tasks['System'].str.upper() == \"LIPIDS METABOLISM\"].copy()\n",
    "lipid_task_names = set(lipid_tasks['Task'].str.strip())\n",
    "\n",
    "nipp_df = enterocyte_df[\n",
    "    (enterocyte_df['condition'] == \"Nippostrongylus\") &\n",
    "    (enterocyte_df['metabolic_task'].isin(lipid_task_names))\n",
    "].copy()\n",
    "\n",
    "# Merge with task_info to get full metadata\n",
    "nipp_df = nipp_df.merge(\n",
    "    adata.metabolic_tasks.uns['task_info'],\n",
    "    how='left',\n",
    "    left_on='metabolic_task',\n",
    "    right_on='Task'\n",
    ")\n",
    "\n",
    "# Sort by scaled_trimean (i.e., enrichment level)\n",
    "nipp_df_sorted = nipp_df.sort_values(\"scaled_trimean\", ascending=False)\n",
    "\n",
    "# Select top 2\n",
    "top2_lipid_tasks = nipp_df_sorted.head(5)\n",
    "\n",
    "\n",
    "print(top2_lipid_tasks[['metabolic_task', 'scaled_trimean', 'System', 'Subsystem']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d48fab53-757a-4291-8639-f61f3ebbeefd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Graph the total profile of each infection (All cells) - potentially biased for cell type abundances\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Create group label for disease x celltype\n",
    "\n",
    "adata.metabolic_tasks.obs['group_label'] = adata.metabolic_tasks.obs['InfectionStatus']\n",
    "\n",
    "agg = sccellfie.expression.aggregation.agg_expression_cells(\n",
    "    adata.metabolic_tasks,\n",
    "    groupby='group_label',\n",
    "    agg_func='trimean'\n",
    ")\n",
    "\n",
    "input_df = sccellfie.preprocessing.matrix_utils.min_max_normalization(agg.T, axis=1)\n",
    "df_melted = pd.melt(input_df.reset_index(), id_vars='Task', var_name='group_label', value_name='scaled_trimean')\n",
    "df_melted = df_melted.rename(columns={'Task': 'metabolic_task'})\n",
    "\n",
    "df_melted['metabolic_task'] = df_melted['metabolic_task'].astype(str).str.strip()\n",
    "adata.metabolic_tasks.uns['task_info']['Task'] = adata.metabolic_tasks.uns['task_info']['Task'].astype(str).str.strip()\n",
    "valid_tasks = set(adata.metabolic_tasks.uns['task_info']['Task'])\n",
    "df_melted = df_melted[df_melted['metabolic_task'].isin(valid_tasks)].copy()\n",
    "\n",
    "\n",
    "#  Get unique conditions present for Enterocytes\n",
    "conditions = df_melted['group_label'].unique()\n",
    "df_melted['cell_type'] = 'All'\n",
    "# Setup subplots\n",
    "n = len(conditions)\n",
    "ncols = 3\n",
    "nrows = (n // ncols) + int(n % ncols != 0)\n",
    "fig, axes = plt.subplots(nrows=nrows, ncols=ncols, subplot_kw={'projection': 'polar'}, figsize=(5 * ncols, 5 * nrows))\n",
    "axes = axes.flatten()\n",
    "\n",
    "# Plot each condition's radial plot\n",
    "for i, (condition, ax) in enumerate(zip(conditions, axes)):\n",
    "    condition_df = df_melted[df_melted['group_label'] == condition]\n",
    "    sccellfie.plotting.create_radial_plot(\n",
    "        condition_df,\n",
    "        adata.metabolic_tasks.uns['task_info'],\n",
    "        ax=ax,\n",
    "        show_legend=(i == 7),\n",
    "        ylim=1.0,\n",
    "        sort_by_value=True\n",
    "    )\n",
    "    ax.set_title(condition)\n",
    "\n",
    "# Remove unused axes\n",
    "for j in range(i + 1, len(axes)):\n",
    "    fig.delaxes(axes[j])\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92d5c81e-11ed-4bd1-9211-04f5d2ec9952",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Create group label for disease x celltype\n",
    "\n",
    "adata.metabolic_tasks.obs['group_label'] = (\n",
    "    adata.metabolic_tasks.obs['secondlevel'].astype(str) + \"_\" +\n",
    "    adata.metabolic_tasks.obs['InfectionStatus'].astype(str)\n",
    ")\n",
    "\n",
    "# Aggregate using group_label column name\n",
    "agg = sccellfie.expression.aggregation.agg_expression_cells(\n",
    "    adata.metabolic_tasks,\n",
    "    groupby='group_label',  # group by this obs column\n",
    "    agg_func='trimean'\n",
    ")\n",
    "\n",
    "# Normalize and melt for plotting\n",
    "input_df = sccellfie.preprocessing.matrix_utils.min_max_normalization(agg.T, axis=1)\n",
    "df_melted = pd.melt(input_df.reset_index(), id_vars='Task', var_name='group_label', value_name='scaled_trimean')\n",
    "df_melted = df_melted.rename(columns={'Task': 'metabolic_task'})\n",
    "# Make sure both task columns are strings and stripped\n",
    "df_melted['metabolic_task'] = df_melted['metabolic_task'].astype(str).str.strip()\n",
    "adata.metabolic_tasks.uns['task_info']['Task'] = adata.metabolic_tasks.uns['task_info']['Task'].astype(str).str.strip()\n",
    "\n",
    "# Filter df_melted to only valid tasks present in task_info\n",
    "valid_tasks = set(adata.metabolic_tasks.uns['task_info']['Task'])\n",
    "df_melted = df_melted[df_melted['metabolic_task'].isin(valid_tasks)].copy()\n",
    "\n",
    "# Filter to only enterocytes (group_label starting with \"Enterocytes_\")\n",
    "enterocyte_df = df_melted[df_melted['group_label'].str.startswith(\"Enterocytes_\")].copy()\n",
    "\n",
    "# Extract condition from group_label for easier filtering and plotting\n",
    "enterocyte_df['condition'] = enterocyte_df['group_label'].str.split('_').str[1]\n",
    "enterocyte_df['cell_type'] = enterocyte_df['group_label'].str.split('_').str[0]\n",
    "\n",
    "conditions = enterocyte_df['condition'].unique()\n",
    "\n",
    "\n",
    "n = len(conditions)\n",
    "ncols = 3\n",
    "nrows = (n // ncols) + int(n % ncols != 0)\n",
    "fig, axes = plt.subplots(nrows=nrows, ncols=ncols, subplot_kw={'projection': 'polar'}, figsize=(5 * ncols, 5 * nrows))\n",
    "axes = axes.flatten()\n",
    "\n",
    "# Plot each condition's radial plot\n",
    "for i, (condition, ax) in enumerate(zip(conditions, axes)):\n",
    "    condition_df = enterocyte_df[enterocyte_df['condition'] == condition]\n",
    "    sccellfie.plotting.create_radial_plot(\n",
    "        condition_df,\n",
    "        adata.metabolic_tasks.uns['task_info'],\n",
    "        ax=ax,\n",
    "        show_legend=(i == 0),\n",
    "        ylim=1.0,\n",
    "        sort_by_value=True\n",
    "    )\n",
    "    ax.set_title(condition)\n",
    "\n",
    "# Remove any unused subplots\n",
    "for j in range(i + 1, len(axes)):\n",
    "    fig.delaxes(axes[j])\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "449aa943-64c1-4b63-9dec-22318f3a99e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Identify Markers - There are Two ways \n",
    "#scCellFie runs a TF-IDF approach, as implemented in the tool SoupX in R. We define a express_cut=5*np.log(2) to define cells expressing each metabolic task. Here, a value of 5*np.log(2) indicates that all reactions in the task are over the threshold value of the determinant gene (or “active”). This condition is often rare in single-cell data due to data sparsity (rarely all reactions in a task are over their determinant gene’s threshold), but it is a good approach for identifying markers.\n",
    "mrks = sccellfie.external.quick_markers(adata.metabolic_tasks,\n",
    "                                        cluster_key=cell_group,\n",
    "                                        n_markers=20,\n",
    "                                        express_cut=5*np.log(2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4024065b-592b-4891-a32f-2f1100f8a08c",
   "metadata": {},
   "outputs": [],
   "source": [
    "mrks.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "407ef7f2-a1c5-4a5c-9bc4-0b02a4899809",
   "metadata": {},
   "outputs": [],
   "source": [
    "#By looking at the distributions given each task TF-IDF scores, we can do some filtering to ensure more specific tasks.\n",
    "mrks['tf_idf'].hist() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d27d8bb3-64ab-45f5-91f1-86cf39cb7ca0",
   "metadata": {},
   "outputs": [],
   "source": [
    "scatter = plt.scatter(mrks['tf'], mrks['idf'], c=mrks['tf_idf'], cmap='YlGnBu')\n",
    "plt.colorbar(scatter)\n",
    "plt.xlabel('TF')\n",
    "plt.ylabel('IDF')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29364d57-efde-426c-a8ff-6c6ebe817e72",
   "metadata": {},
   "outputs": [],
   "source": [
    "#scCellFie includes a filtering function that fits a hyperbolic curve, while the user can define some thresholds based on the tf-idf scores, and the ratio between the top and second best cell clusters (tf_ratio)\n",
    "x_col = 'tf'\n",
    "y_col = 'idf'\n",
    "df = mrks\n",
    "tfidf_threshold = 0.3\n",
    "tf_ratio = 1.2\n",
    "\n",
    "# Visualization\n",
    "plt.figure(figsize=(6, 6))\n",
    "\n",
    "# Plot all points\n",
    "plt.scatter(df[x_col], df[y_col], alpha=0.5, c='lightgray', label='All points')\n",
    "\n",
    "# Plot selected points\n",
    "filtered_mrks, curve = sccellfie.external.filter_tfidf_markers(df, tf_col=x_col, idf_col=y_col, tfidf_threshold=tfidf_threshold, tf_ratio=tf_ratio)\n",
    "\n",
    "plt.scatter(filtered_mrks[x_col], filtered_mrks[y_col], c='red', label='Selected points')\n",
    "plt.plot(*curve, label='Fit curve')\n",
    "\n",
    "plt.xlabel('TF')\n",
    "plt.ylabel('IDF')\n",
    "plt.legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0d85622-2b3d-46ca-80e7-685e23b28711",
   "metadata": {},
   "outputs": [],
   "source": [
    "tf_idf_mrks = filtered_mrks['gene'].unique().tolist()\n",
    "len(tf_idf_mrks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7624eaf5-346a-4987-ab2b-fad67cbae56b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#We can plot the defining metabolic tasks?\n",
    "plt.rcParams['figure.figsize'] = (3,3)\n",
    "plt.rcParams['font.size'] = 10\n",
    "\n",
    "sc.pl.embedding(adata.metabolic_tasks,\n",
    "                color=[cell_group] + tf_idf_mrks,\n",
    "                ncols=3,\n",
    "                palette=palette,\n",
    "                frameon=False,\n",
    "                basis='X_wnn.umap_cc',\n",
    "                wspace=0.7,\n",
    "                title=[\"\\n\".join(textwrap.wrap(t, width=60)) for t in [cell_group] + tf_idf_mrks],\n",
    "                cmap='OrRd'\n",
    "               )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "478e0d04-ea77-4738-ae1f-7d31701ac047",
   "metadata": {},
   "outputs": [],
   "source": [
    "#A different method to identify metabolism marker pathways\n",
    "#Similarly, we can use another approach for identifying metabolic markers. In this case, a logistic regression implemented in Scanpy.\n",
    "method = 'logreg'\n",
    "sc.tl.rank_genes_groups(adata.metabolic_tasks, cell_group, method=method,\n",
    "                        use_raw=False, key_added = method)\n",
    "\n",
    "sc.pl.rank_genes_groups(adata.metabolic_tasks, n_genes=20, sharey=True, key=method, fontsize=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a6f8ee8-8e9a-49b0-9f0f-8aaa268967db",
   "metadata": {},
   "outputs": [],
   "source": [
    "scanpy_df = sc.get.rank_genes_groups_df(adata.metabolic_tasks,\n",
    "                                          key=method,\n",
    "                                          group=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "584fcfaa-c8ea-4855-be64-e6f3d6d9f6db",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.rank_genes_groups_dotplot(adata.metabolic_tasks, n_genes=1, groupby=cell_group,\n",
    "                                key=method, use_raw=False, standard_scale='var',\n",
    "                               )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62397ab1-519f-4b1b-bbcd-c9883d12025d",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_group2 = \"FinestCellType\"\n",
    "\n",
    "sc.pl.rank_genes_groups_dotplot(adata.metabolic_tasks, n_genes=1, groupby=cell_group2,\n",
    "                                key=method, use_raw=False, standard_scale='var',\n",
    "                               )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aecc32b5-7102-49d9-b467-e5ab090f3965",
   "metadata": {},
   "outputs": [],
   "source": [
    "#We can perform Differential expression analysis \n",
    "adata.obs.InfectionStatus.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b18c4a89-4d11-41ce-b6ff-bcee8f688d91",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Diff regulation part 2 \n",
    "condition_key = 'InfectionStatus'\n",
    "reference = 'Naive'\n",
    "\n",
    "# Get list of experimental conditions excluding \"Naive\"\n",
    "experimental_conditions = adata.obs[condition_key].cat.categories\n",
    "experimental_conditions = [cond for cond in experimental_conditions if cond != reference]\n",
    "\n",
    "# Create contrast pairs \n",
    "contrasts = [(reference, cond) for cond in experimental_conditions]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a7b4009-2e5a-4144-9e46-e6e217a1ca8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "all_results = []\n",
    "\n",
    "for cell_type in adata.obs[cell_group].unique():\n",
    "    for cond1, cond2 in contrasts:\n",
    "        dma = sccellfie.stats.scanpy_differential_analysis(adata.metabolic_tasks,\n",
    "                                                           cell_type=cell_type,\n",
    "                                                           cell_type_key=cell_group,\n",
    "                                                           condition_key=condition_key,\n",
    "                                                           min_cells=20,\n",
    "                                                           condition_pairs=[(cond1, cond2)])\n",
    "\n",
    "        if dma is not None and not dma.empty:\n",
    "            dma['cell_type'] = cell_type\n",
    "            dma['contrast'] = f\"{cond2}_vs_{cond1}\"\n",
    "            all_results.append(dma)\n",
    "\n",
    "# Combine into single dataframe\n",
    "dma_combined = pd.concat(all_results, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85b7a395-bd59-4f57-8260-e4356501ad7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Adjust P values globally for the multiple corrections\n",
    "dma_combined['pval_adj'] = multipletests(dma_combined['p_value'], method='fdr_bh')[1]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a6c4420-aabf-47a0-9c32-936c7ece86a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(dma_combined.columns.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7db7dfd2-423b-4de0-a491-7f72de37acac",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Filter significant results \n",
    "# Thresholds\n",
    "pval_threshold = 0.05\n",
    "cohen_threshold = .5\n",
    "logfc_threshold = 0.585 #is equal to 1.5 FC\n",
    "\n",
    "sig_results = dma_combined[\n",
    "    (dma_combined['pval_adj'] < pval_threshold) &\n",
    "    (dma_combined['cohens_d'].abs() > cohen_threshold) &\n",
    "    (dma_combined['log2FC'].abs() > logfc_threshold)\n",
    "]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59354ba9-e985-4a90-abb6-31cd9a90253e",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(sig_results.columns.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab638dfd-4995-4ccd-86a6-260fab013f9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Target cell types\n",
    "target_cells = [\"Enterocytes\", \"Stem Cells\", \"Transit Amplifying Cells\", \"ILC2s\"]\n",
    "\n",
    "# Filter for significant features in target cells vs Naive\n",
    "sig_naive = sig_results[\n",
    "    (sig_results['cell_type'].isin(target_cells)) \n",
    "]\n",
    "\n",
    "# Select relevant columns and drop duplicates if any\n",
    "sig_table = sig_naive[['feature', 'contrast', 'cell_type']].drop_duplicates()\n",
    "\n",
    "# Show the table (or print, or save)\n",
    "print(sig_table)\n",
    "\n",
    "# Optionally, group by condition/cell type to see counts\n",
    "counts = sig_table.groupby(['contrast', 'cell_type']).size().reset_index(name='n_pathways')\n",
    "print(counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b45843c-bf93-4757-a0b0-1e44d41a3ec1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Visualize how many differentially expressed pathways we observe\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "# Assume counts_full from before (with all combos and zero-filled)\n",
    "# If you don't have counts_full yet, create it as:\n",
    "\n",
    "all_contrasts = sig_table['contrast'].unique()\n",
    "all_cells = sig_table['cell_type'].unique()\n",
    "\n",
    "full_index = pd.MultiIndex.from_product([all_cells, all_contrasts], names=['cell_type', 'contrast'])\n",
    "counts = sig_table.groupby(['cell_type', 'contrast']).size().reset_index(name='n_pathways')\n",
    "counts_full = counts.set_index(['cell_type', 'contrast']).reindex(full_index, fill_value=0).reset_index()\n",
    "\n",
    "# Plot \n",
    "plt.figure(figsize=(12,6))\n",
    "sns.barplot(data=counts_full, x='cell_type', y='n_pathways', hue='contrast', palette='tab10')\n",
    "\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.ylabel('Number of Significant Pathways')\n",
    "plt.xlabel('Cell Type')\n",
    "plt.title('Significant Metabolic Pathways by Cell Type and Condition')\n",
    "plt.legend(title='Condition', bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6dc9a33a-6828-4a1c-b236-833b8cc2246f",
   "metadata": {},
   "outputs": [],
   "source": [
    "contrast_of_interest = \"Nippostrongylus_vs_Naive\"\n",
    "cell_of_interest = \"Enterocytes\"\n",
    "\n",
    "# Filter significant features based on your thresholds\n",
    "sig = dma_combined[\n",
    "    (dma_combined['contrast'] == contrast_of_interest) &\n",
    "    (dma_combined['cell_type'] == cell_of_interest) &\n",
    "    (dma_combined['pval_adj'] < pval_threshold) &\n",
    "    (dma_combined['cohens_d'].abs() > cohen_threshold) &\n",
    "    (dma_combined['log2FC'].abs() > logfc_threshold)\n",
    "].set_index('feature')\n",
    "\n",
    "sig_features = sig.sort_values('log2FC').index.tolist()\n",
    "\n",
    "print(f\"Found {len(sig_features)} significant metabolic tasks for {cell_of_interest} in {contrast_of_interest}\")\n",
    "print(sig_features)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3c5e088-393f-488c-8d99-b67b05c8c8d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "dma_combined.head\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24b12a1c-1e6b-421a-a5c5-dadec3d0908f",
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(sig_features) > 0:\n",
    "    width = min(18, 1.5 + 0.5 * len(sig_features))\n",
    "    fig, ax = sccellfie.plotting.create_comparative_violin(\n",
    "        adata=adata.metabolic_tasks,\n",
    "        significant_features=sig_features,\n",
    "        group1='Naive',\n",
    "        group2='Nippostrongylus',\n",
    "        condition_key=condition_key,      \n",
    "        cell_type_key=cell_group,        \n",
    "        celltype=cell_of_interest,\n",
    "        xlabel='',\n",
    "        ylabel='Metabolic Score',\n",
    "        palette=['gray', 'darkgreen'],   \n",
    "        wrapped_title_length=100,\n",
    "        figsize=(width, 4),\n",
    "        fontsize=14,\n",
    "        title=f'{cell_of_interest} ({contrast_of_interest})',\n",
    "        tight_layout=False,\n",
    "        # save=f'Violin-{cell_of_interest}-{contrast_of_interest}.pdf'  \n",
    "    )\n",
    "else:\n",
    "    print(f\"No significant features found for {cell_of_interest} in {contrast_of_interest} under current thresholds.\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd2f0c97-0720-4083-944b-868e08a07868",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot a subset Lipis metabolism results for Nippo for the paper \n",
    "import matplotlib as mpl\n",
    "mpl.rcParams['svg.fonttype'] = 'none' \n",
    "\n",
    "sig_features2 = [\n",
    "    \"Acetoacetate synthesis\",\n",
    "    \"Arachidonate degradation\",\n",
    "    \"Arachidonate synthesis\", \"Ceramide synthesis\", \"Linoleate degradation\"\n",
    "]\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "\n",
    "\n",
    "# Subset the adata to only Enterocytes and selected tasks\n",
    "# Filter for relevant cells and conditions\n",
    "cell_mask = adata.metabolic_tasks.obs[cell_group] == cell_of_interest\n",
    "condition_mask = adata.metabolic_tasks.obs[condition_key].isin(['Naive', 'Nippostrongylus'])\n",
    "\n",
    "adata_subset = adata.metabolic_tasks[cell_mask & condition_mask, :]\n",
    "\n",
    "# Extract scores for selected tasks\n",
    "scores = pd.DataFrame(adata_subset[:, sig_features2].X.toarray(), \n",
    "                      columns=sig_features2,\n",
    "                      index=adata_subset.obs_names)\n",
    "\n",
    "# Add metadata\n",
    "scores['condition'] = adata_subset.obs[condition_key].values\n",
    "scores['cell_id'] = scores.index\n",
    "\n",
    "# Convert to long format\n",
    "df_long = pd.melt(scores,\n",
    "                  id_vars=['cell_id', 'condition'],\n",
    "                  var_name='Metabolic Task',\n",
    "                  value_name='Score')\n",
    "\n",
    "# Plot\n",
    "plt.figure(figsize=(7, 5))\n",
    "sns.violinplot(\n",
    "    data=df_long,\n",
    "    x='Score',\n",
    "    y='Metabolic Task',\n",
    "    hue='condition',\n",
    "    split=True,\n",
    "    inner='quartile',\n",
    "    palette={'Naive': 'lightgrey', 'Nippostrongylus': 'navy'}\n",
    ")\n",
    "\n",
    "plt.title(f\"{cell_of_interest}: Metabolic Task Activity (Split Violin)\")\n",
    "plt.xlabel(\"Metabolic Score\")\n",
    "plt.ylabel(\"\")\n",
    "plt.xticks(rotation=30, ha='right')\n",
    "plt.legend(title=\"Condition\", bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig(\"/path/to/analysis/directory/scCellfie/Images/split_violin_metabolic_tasks.svg\", format=\"svg\", dpi=300)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "989003de-584d-44b9-bab1-42d861e77a99",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib as mpl\n",
    "mpl.rcParams['svg.fonttype'] = 'none'\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "\n",
    "sig_features2 = [\n",
    "    \"Acetoacetate synthesis\",\n",
    "    \"Arachidonate degradation\",\n",
    "    \"Arachidonate synthesis\",\n",
    "    \"Ceramide synthesis\",\n",
    "    \"Linoleate degradation\"\n",
    "]\n",
    "\n",
    "# Subset the adata\n",
    "cell_mask = adata.metabolic_tasks.obs[cell_group] == cell_of_interest\n",
    "condition_mask = adata.metabolic_tasks.obs[condition_key].isin(['Naive', 'Nippostrongylus'])\n",
    "adata_subset = adata.metabolic_tasks[cell_mask & condition_mask, :]\n",
    "\n",
    "# Extract scores\n",
    "scores = pd.DataFrame(\n",
    "    adata_subset[:, sig_features2].X.toarray(), \n",
    "    columns=sig_features2,\n",
    "    index=adata_subset.obs_names\n",
    ")\n",
    "\n",
    "# Add metadata\n",
    "scores['condition'] = adata_subset.obs[condition_key].values\n",
    "scores['cell_id'] = scores.index\n",
    "\n",
    "# Convert to long format\n",
    "df_long = pd.melt(\n",
    "    scores,\n",
    "    id_vars=['cell_id', 'condition'],\n",
    "    var_name='Metabolic Task',\n",
    "    value_name='Score'\n",
    ")\n",
    "\n",
    "# Plot\n",
    "plt.figure(figsize=(8, 5))\n",
    "sns.violinplot(\n",
    "    data=df_long,\n",
    "    x='Score',\n",
    "    y='Metabolic Task',\n",
    "    hue='condition',\n",
    "    inner='quartile',\n",
    "    palette={'Naive': 'lightgrey', 'Nippostrongylus': 'navy'},\n",
    "    dodge=True  # default is True, can be explicit\n",
    ")\n",
    "\n",
    "plt.title(f\"{cell_of_interest}: Metabolic Task Activity\")\n",
    "plt.xlabel(\"Metabolic Score\")\n",
    "plt.ylabel(\"\")\n",
    "plt.xticks(rotation=30, ha='right')\n",
    "plt.legend(title=\"Condition\", bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig(\"/path/to/analysis/directory/scCellfie/Images/split_violin_metabolic_tasks.svg\", format=\"svg\", dpi=300)\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ada494c-fa91-456a-acf0-ea32321a53fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "contrast_of_interest = \"Nippostrongylus_vs_Naive\"\n",
    "cell_of_interest = \"Enterocytes\"\n",
    "\n",
    "# Step 1: Get all tasks in the Lipids Metabolism system\n",
    "task_info = adata.metabolic_tasks.uns['task_info']\n",
    "lipid_tasks = task_info[task_info['System'].str.upper() == 'LIPIDS METABOLISM']['Task'].str.strip().tolist()\n",
    "lipid_tasks_set = set(lipid_tasks)\n",
    "\n",
    "# Step 2: Intersect with significant features from differential analysis\n",
    "sig_lipid_features = [f for f in sig_features if f in lipid_tasks_set]\n",
    "\n",
    "print(f\"Found {len(sig_lipid_features)} significant Lipids Metabolism tasks for {cell_of_interest} in {contrast_of_interest}\")\n",
    "print(sig_lipid_features)\n",
    "\n",
    "# Proceed if there are any significant lipid-related tasks\n",
    "if len(sig_lipid_features) > 0:\n",
    "    # Step 3: Subset the adata to just Enterocytes under Naive or Nippostrongylus\n",
    "    cell_mask = adata.metabolic_tasks.obs[cell_group] == cell_of_interest\n",
    "    condition_mask = adata.metabolic_tasks.obs[condition_key].isin(['Naive', 'Nippostrongylus'])\n",
    "    adata_subset = adata.metabolic_tasks[cell_mask & condition_mask, :]\n",
    "\n",
    "    # Extract scores for the lipid tasks\n",
    "    scores = pd.DataFrame(\n",
    "        adata_subset[:, sig_lipid_features].X.toarray(),\n",
    "        columns=sig_lipid_features,\n",
    "        index=adata_subset.obs_names\n",
    "    )\n",
    "\n",
    "    # Add metadata\n",
    "    scores['condition'] = adata_subset.obs[condition_key].values\n",
    "    scores['cell_id'] = scores.index\n",
    "\n",
    "    # Step 6: Melt to long format\n",
    "    df_long = pd.melt(\n",
    "        scores,\n",
    "        id_vars=['cell_id', 'condition'],\n",
    "        var_name='Metabolic Task',\n",
    "        value_name='Score'\n",
    "    )\n",
    "\n",
    "\n",
    "    mpl.rcParams['svg.fonttype'] = 'none'\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=(min(2 + 0.4 * len(sig_lipid_features), 24), 3.0))\n",
    "    sns.violinplot(\n",
    "    data=df_long,\n",
    "    x='Metabolic Task',   # ⬅️ now on x-axis\n",
    "    y='Score',            # ⬅️ now on y-axis\n",
    "    hue='condition',\n",
    "    inner='box',\n",
    "    palette={'Naive': 'lightgrey', 'Nippostrongylus': '#ECB364'},\n",
    "    dodge=True,\n",
    "    linewidth=1,\n",
    "    width=0.8,\n",
    "    ax=ax\n",
    "    )\n",
    "\n",
    "    for collection in ax.collections:\n",
    "        if isinstance(collection, mpl.collections.PolyCollection):\n",
    "            collection.set_edgecolor('black')\n",
    "            collection.set_linewidth(1)\n",
    "        \n",
    "    #plt.title(f\"{cell_of_interest}: Lipids Metabolism Task Activity\\n{contrast_of_interest}\")\n",
    "    plt.xlabel(\"\")\n",
    "    plt.ylabel(\"Metabolic Score\")\n",
    "    plt.xticks(rotation=30, ha='right')\n",
    "    plt.legend(title=\"Condition\", bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(\"/path/to/analysis/directory/scCellfie/Images/split_violin_metabolic_tasks_Nippo_lipids_Sig.svg\", format=\"svg\", dpi=300)\n",
    "    plt.show()\n",
    "\n",
    "else:\n",
    "    print(\"No significant lipid metabolism tasks found under current thresholds.\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0596f5f0-9e84-4e42-88af-e3dea42fb1ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(min(2 + 0.6 * len(sig_lipid_features), 30), 3.0))  # Wider, not taller\n",
    "\n",
    "flierprops = dict(marker='o', markersize=2, markerfacecolor='none', markeredgecolor='black', markeredgewidth=0.5)\n",
    "\n",
    "sns.boxplot(\n",
    "    data=df_long,\n",
    "    x='Metabolic Task',\n",
    "    y='Score',\n",
    "    hue='condition',\n",
    "    palette={'Naive': 'lightgrey', 'Nippostrongylus': '#ECB364'},\n",
    "    dodge=True,\n",
    "    linewidth=1,\n",
    "    width=0.8,\n",
    "    flierprops=flierprops,\n",
    "    ax=ax\n",
    ")\n",
    "\n",
    "# Force black outlines on boxes and whiskers\n",
    "for line in ax.lines:\n",
    "    line.set_color('black')\n",
    "    line.set_linewidth(1)\n",
    "\n",
    "for patch in ax.artists:\n",
    "    patch.set_edgecolor('black')\n",
    "    patch.set_linewidth(1)\n",
    "\n",
    "# Add grid lines\n",
    "ax.yaxis.grid(True, which='major', linestyle='--', linewidth=0.5, color='gray')\n",
    "ax.xaxis.grid(False)\n",
    "\n",
    "# Axis labels and formatting\n",
    "plt.xlabel(\"\")\n",
    "plt.ylabel(\"Metabolic Score\")\n",
    "plt.xticks(rotation=30, ha='right')\n",
    "plt.legend(title=\"Condition\", bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "plt.tight_layout()\n",
    "\n",
    "# Save the plot\n",
    "#Figure 5B\n",
    "plt.savefig(\n",
    "    \"/path/to/analysis/directory/scCellfie/Images/split_boxplot_metabolic_tasks_Nippo_lipids_Sig.svg\",\n",
    "    format=\"svg\", dpi=300\n",
    ")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b52b3f70-6a00-4f44-ac73-af5054dca09e",
   "metadata": {},
   "outputs": [],
   "source": [
    "dma_filtered = dma_combined[\n",
    "    (dma_combined['contrast'] == contrast_of_interest) &\n",
    "    (dma_combined['cell_type'] == cell_of_interest)\n",
    "]\n",
    "\n",
    "# Filter to only the lipid metabolism tasks that you plotted\n",
    "dma_plot_tasks = dma_filtered[dma_filtered['feature'].isin(sig_lipid_features)]\n",
    "\n",
    "# Select and sort relevant columns\n",
    "logfc_table = dma_plot_tasks[['feature', 'log2FC', 'pval_adj']].sort_values('log2FC', ascending=False)\n",
    "\n",
    "print(logfc_table)\n",
    "print(dma_combined)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b3df025-8c33-4b2e-88d5-5a9e97377c3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create a Heatmap of selected tasks to ask if these tasks are broadly upregulated in cell types during Nippo infection\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "\n",
    "# Define custom colormap\n",
    "custom_cmap = LinearSegmentedColormap.from_list(\n",
    "    \"navy_to_red\", [\"navy\", \"lightgrey\", \"darkred\"]\n",
    ")\n",
    "\n",
    "# Define task list\n",
    "tasks_of_interest = [\n",
    "    'Arachidonate synthesis', \n",
    "    'Palmitate degradation', \n",
    "    'Elaidate degradation', \n",
    "    'cis-vaccenic acid degradation', \n",
    "    'Linoleate degradation'\n",
    "]\n",
    "\n",
    "# Filter for contrast and tasks of interest across all cell types\n",
    "subset = dma_combined[\n",
    "    (dma_combined['contrast'] == \"Nippostrongylus_vs_Naive\") &\n",
    "    (dma_combined['feature'].isin(tasks_of_interest))\n",
    "]\n",
    "\n",
    "# Pivot to heatmap format: tasks (rows) x cell types (columns)\n",
    "heatmap_data = subset.pivot(index='feature', columns='cell_type', values='median_diff')\n",
    "\n",
    "# Optional: enforce row order\n",
    "heatmap_data = heatmap_data.loc[tasks_of_interest]\n",
    "\n",
    "# Plot heatmap\n",
    "plt.figure(figsize=(10, 3))\n",
    "sns.heatmap(\n",
    "    heatmap_data,\n",
    "    cmap=custom_cmap,\n",
    "    center=0,\n",
    "    linewidths=0.5,\n",
    "    linecolor='grey',\n",
    "    annot=False,\n",
    "    fmt=\".2f\",\n",
    "    cbar_kws={\"label\": \"Median Diff\"}\n",
    ")\n",
    "plt.xlabel(None)\n",
    "plt.ylabel(None)\n",
    "#plt.title(\"Lipid Task Activity (Log2FC) across Cell Types\")\n",
    "plt.tight_layout()\n",
    "plt.savefig(\n",
    "    \"/path/to/analysis/directory/scCellfie/Images/Heatmap_metabolic_tasks_Nippo_lipids_Sig.svg\",\n",
    "    format=\"svg\", dpi=300\n",
    ")\n",
    "plt.show()\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08f75508-e12c-40dd-a1b8-9630a085324d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add the finecell type data to the differntial analysis results by creating a mapping\n",
    "coarse_to_fine = (\n",
    "    adata.obs.groupby(cell_group)[\"FineCellType\"]\n",
    "    .apply(lambda x: sorted(x.unique()))\n",
    ")\n",
    "\n",
    "expanded_results = []\n",
    "\n",
    "for _, row in dma_combined.iterrows():\n",
    "    coarse_type = row[\"cell_type\"]\n",
    "\n",
    "    # If it's Enterocytes, split to fine cell types\n",
    "    if coarse_type == \"Enterocytes\":\n",
    "        fine_types = coarse_to_fine.get(coarse_type, [None])\n",
    "        for fine in fine_types:\n",
    "            new_row = row.copy()\n",
    "            new_row[\"fine_cell_type\"] = fine\n",
    "            new_row[\"cell_type_plot\"] = f\"{fine}\"  # label x-axis by fine type only\n",
    "            expanded_results.append(new_row)\n",
    "\n",
    "    # Otherwise, keep as single coarse type\n",
    "    else:\n",
    "        new_row = row.copy()\n",
    "        new_row[\"fine_cell_type\"] = coarse_type  # keep same label\n",
    "        new_row[\"cell_type_plot\"] = coarse_type\n",
    "        expanded_results.append(new_row)\n",
    "\n",
    "dma_expanded = pd.DataFrame(expanded_results)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39cbfb5c-e09b-43cc-872a-6c1257d11eeb",
   "metadata": {},
   "outputs": [],
   "source": [
    "for ct in dma_expanded[\"cell_type_plot\"].unique():\n",
    "    print(ct)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8975ae27-c2a8-4a2f-8b91-e8726aa6eeec",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(dma_expanded)\n",
    "\n",
    "desired_order = [\n",
    "     \"Early Enterocyte\", \"Middle Enterocyte\", \"Late Enterocyte\",\n",
    "    \"Goblet Cells\",\"Enteroendocrine Cells\",  \"Paneth Cells\",\n",
    "    \"Transit Amplifying Cells\", \"Tuft Cells\", \"Stem Cells\",\n",
    "    \"CD8 ab T Cells\", \"Effector CD8 ab T Cells\",\n",
    "    \"gd T Cells\", \"Cycling T Cells\", \"CD4 ab T Cells\",\n",
    "    \"Effector CD4 ab T Cells\", \"NKT Cells\", \"NK Cells\",\n",
    "    \"ILC1s\", \"ILC2s\", \"ILC3s\", \n",
    "    \"B Cells\", \"GC B Cells\", \"Plasma Cells\",\n",
    "     \"cDC2s\",  \"Macrophages\",\n",
    "    \"Monocytes\",  \"Plasmacytoid DCs\",\n",
    "    \"Lymphatic Endothelial Cells\", \n",
    "    \"Fibroblasts\", \"Smooth Muscle Cells\"\n",
    "]\n",
    "\n",
    "dma_expanded[\"cell_type_plot\"] = pd.Categorical(\n",
    "    dma_expanded[\"cell_type_plot\"],\n",
    "    categories=desired_order,\n",
    "    ordered=True\n",
    ")\n",
    "tasks_of_interest = [\n",
    "    'Arachidonate synthesis', \n",
    "    'Palmitate degradation', \n",
    "    'Elaidate degradation', \n",
    "    'cis-vaccenic acid degradation', \n",
    "    'Linoleate degradation'\n",
    "]\n",
    "subset = dma_expanded[\n",
    "    (dma_expanded['contrast'] == \"Nippostrongylus_vs_Naive\") &\n",
    "    (dma_expanded['feature'].isin(tasks_of_interest))\n",
    "]\n",
    "\n",
    "heatmap_data = subset.pivot(\n",
    "    index='feature',\n",
    "    columns='cell_type_plot',\n",
    "    values='median_diff'\n",
    ")\n",
    "\n",
    "# Reorder columns to match biological order\n",
    "heatmap_data = heatmap_data.reindex(columns=desired_order)\n",
    "\n",
    "\n",
    "plt.figure(figsize=(12, 4.5))\n",
    "sns.heatmap(\n",
    "    heatmap_data,\n",
    "    cmap=custom_cmap,\n",
    "    center=0,\n",
    "    linewidths=0.5,\n",
    "    linecolor='grey',\n",
    "    cbar_kws={\"label\": \"Median Diff\"}\n",
    ")\n",
    "plt.tight_layout()\n",
    "plt.savefig(\n",
    "    \"/path/to/analysis/directory/scCellfie/Images/Heatmap_metabolic_tasks_Nippo_lipids_Sig_EnterocyteSplit.svg\",\n",
    "    format=\"svg\", dpi=300\n",
    ")\n",
    "plt.show()\n",
    "#Figure 5C"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scCEllfie scRNA",
   "language": "python",
   "name": "sccellfie_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
